controllers/virtualmachine/volumebatch/volumebatch_controller.go

```diff
@@ -293,95 +297,87 @@ func (r *Reconciler) processBatchAttachment(
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      attachmentName,
-			Namespace: ctx.VM.Namespace,
-		},
-		Spec: cnsv1alpha1.CnsNodeVmBatchAttachmentSpec{
-			NodeUUID: ctx.VM.Status.BiosUUID,
-			Volumes:  volumeSpecs,
-		},
-	}
+	vm := ctx.VM
+	attachmentName := pkgutil.CNSBatchAttachmentNameForVolume(vm.Name)
+	batchAttachment := &cnsv1alpha1.CnsNodeVmBatchAttachment{}
+	batchAttachment.Name = attachmentName
+	batchAttachment.Namespace = vm.Namespace
 
-	if err := controllerutil.SetControllerReference(ctx.VM, attachment, r.Client.Scheme()); err != nil {
-		return fmt.Errorf("cannot set controller reference on CnsNodeVmBatchAttachment: %w", err)
-	}
-
-	if err := r.Create(ctx, attachment); err != nil {
-		return fmt.Errorf("cannot create CnsNodeVmBatchAttachment: %w", err)
-	}
+	// Use controllerutil.CreateOrPatch
+	// The operationFn will be called if the object doesn't exist (creating)
+	// or if it does exist (patching).
+	operationResult, err := controllerutil.CreateOrPatch(
+		ctx,
+		r.Client,
+		batchAttachment,
+		func() error {
+			if err := controllerutil.SetControllerReference(
+				vm, batchAttachment, r.Client.Scheme(),
+			); err != nil {
+				return fmt.Errorf("failed to set controller reference "+
+					"on CnsNodeVmBatchAttachment: %w", err)
+			}
 
-	ctx.Logger.Info("Created CnsNodeVmBatchAttachment", "attachment", attachmentName)
-	return nil
-}
+			// Update the Spec with the desired volumeSpecs
+			batchAttachment.Spec = cnsv1alpha1.CnsNodeVmBatchAttachmentSpec{
+				NodeUUID: vm.Status.BiosUUID, // Assuming vm.Status.BiosUUID is accessible
+				Volumes:  volumeSpecs,
+			}
 
-func (r *Reconciler) updateBatchAttachment(
-	ctx *pkgctx.VolumeContext,
-	existing *cnsv1alpha1.CnsNodeVmBatchAttachment,
-	volumes []vmopv1.VirtualMachineVolume) error {
+			return nil
+		})
 
-	volumeSpecs, err := r.buildVolumeSpecs(volumes)
 	if err != nil {
-		return fmt.Errorf("failed to build volume specs: %w", err)
+		return fmt.Errorf("failed to create or patch CnsNodeVmBatchAttachment %s: %w",
+			attachmentName, err)
 	}
 
-	// TODO (Oracle RAC): this should not just be called when the spec differs.
-	// Check if update is needed.
-	if r.volumeSpecsEqual(existing.Spec.Volumes, volumeSpecs) {
-		// No update needed, just update VM status from existing attachment
-		return r.updateVMStatusFromBatchAttachment(ctx, existing)
+	switch operationResult {
+	case controllerutil.OperationResultCreated:
+		ctx.Logger.Info("Created CnsNodeVmBatchAttachment",
+			"attachment", attachmentName)
+	case controllerutil.OperationResultUpdated:
+		ctx.Logger.Info("Updated CnsNodeVmBatchAttachment",
+			"attachment", attachmentName)
 	}
 
-	// Update the batch attachment with new volume specs
-	existing.Spec.Volumes = volumeSpecs
-	if err := r.Update(ctx, existing); err != nil {
-		return fmt.Errorf("cannot update CnsNodeVmBatchAttachment: %w", err)
-	}
+	// Update VM.status.volumes after BatchAttachment is updated.
```

Comment:
return early if create?

Reply:
No, for create also need to create empty vm volume status